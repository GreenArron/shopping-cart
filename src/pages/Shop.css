main.shop {
  width: 100%;
  display: flex;
  flex-direction: column;
  padding: 24px 62px;
}

main.shop section.controls {
  width: 100%;
  padding: 0 0.5rem;
  margin: 0.2rem 0 1rem;
  display: flex;
  justify-content: space-between;
}

main.shop .sort-container {
  display: flex;
  flex-direction: row;
  gap: 8px;
}

main.shop .sort-container .sortorder {
  align-self: center;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 2rem;
  height: 2rem;
  padding: 0;

  background-color: white;
  border: 1px solid black;
  border-radius: 70%;

  transition: transform 120ms ease-in-out;
}

main.shop .sort-container .sortorder.up {
  transform: rotate(180deg);
}

main.shop .sort-container .sortorder svg {
  width: 2em;
}

main.shop section.products {
  width: 100%;
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  gap: 12px;
}

main.shop .shop-product {
  display: flex;
  flex-direction: column;
  align-items: center;

  margin-top: 4px;

  width: 26rem;
  height: auto;
  font-weight: 600;
  padding: 16px 8px;
  background-color: white;
  border-radius: 8px;
}

main.shop .shop-product img {
  aspect-ratio: 1 / 1;
  max-width: 22rem;
  height: auto;
  object-fit: contain;
}

main.shop .shop-product .product-info {
  margin-top: 28px;
  display: flex;
  flex-direction: column;
  gap: 4px;
}

main.shop .shop-product .product-name {
  max-width: 20rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  /* h3 stuff */
  font-size: 1.125rem;
  margin: 0.5rem 0;
}

main.shop .shop-product .rating {
  position: relative;
  display: flex;
  justify-content: center;
  font-size: 0.75rem;
  padding-left: 4ch;
}

main.shop .shop-product .price {
  background-color: #5eba5e;
  color: white;
  width: fit-content;
  align-self: center;
  padding: 6px;
  border-radius: 8px;
}

main.shop .shop-product .buy-controls {
  width: 80%;
  margin-top: 4px;
  display: flex;
  justify-content: end;
}

main.shop .shop-product .buy-controls button.cart-add {
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 1.25rem;

  padding: 0.5rem;

  background-color: var(--primary);
}

main.shop .shop-product .buy-controls button.cart-add svg {
  display: flex;
  align-items: center;
  width: 1.25rem;
  margin-right: 4px;
}
