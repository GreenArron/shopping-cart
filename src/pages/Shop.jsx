import { useEffect, useState } from "react";
import PropTypes from "prop-types";
import "./Shop.css";
import { Filter, FilterItem } from "../components/Filter";
import { fetchProducts, fetchCategories } from "../utils/fakestore-api";
import StarRating from "../components/StarRating";
import SelectCount from "../components/SelectCount";
import { useOutletContext } from "react-router-dom";

function Product({ title, image, price, rating, inCartCount, onCountChange }) {
  return (
    <div className="shop-product">
      <img src={image} />
      <div className="product-info">
        <div className="product-name">{title}</div>
        <div className="rating">
          <StarRating rating={rating.rate} total={5} />
          <span className="rate-count">({rating.count})</span>
        </div>
        <div className="price">$ {price}</div>
      </div>
      <div className="buy-controls">
        <SelectCount count={inCartCount} setCount={onCountChange} />
      </div>
    </div>
  );
}

Product.propTypes = {
  id: PropTypes.number,
  title: PropTypes.string,
  image: PropTypes.string,
  price: PropTypes.number,
  category: PropTypes.string,
  inCartCount: PropTypes.number,
  onCountChange: PropTypes.func,
  rating: PropTypes.shape({
    rate: PropTypes.number,
    count: PropTypes.number,
  }),
};

function Shop() {
  // eslint-disable-next-line no-unused-vars
  const [cart, setCart] = useOutletContext();
  const [category, setCategory] = useState(null);
  const [products, setProducts] = useState([]);
  const [categories, setCategories] = useState([]);

  function changeCart(func, productId) {
    const change = func(0);
    const currentCount = cart[productId] === undefined ? 0 : cart[productId];
    const newCount = Math.max(currentCount + change, 0);
    const changed = newCount - currentCount;

    if (changed !== 0) {
      setCart((currCart) => {
        return { ...currCart, [productId]: newCount };
      });
    }
  }

  useEffect(() => {
    fetchCategories().then((allCategories) => setCategories(allCategories));
  }, []);

  useEffect(() => {
    fetchProducts({ category: category }).then((newProducts) =>
      setProducts(newProducts),
    );
  }, [category]);

  return (
    <main className="shop">
      <section className="controls" aria-label="filter controls">
        <Filter name="Filter" />

        <Filter name="Sort By" />
      </section>
      <section className="products" aria-label="shop items">
        {products.map((product) => {
          return (
            <Product
              key={product.id}
              {...product}
              onCountChange={(func) => changeCart(func, product.id)}
              inCartCount={
                cart[product.id] === undefined ? 0 : cart[product.id]
              }
            />
          );
        })}
      </section>
    </main>
  );
}

Shop.propTypes = {
  cartCount: PropTypes.number,
  setCartCount: PropTypes.func,
};
export default Shop;
