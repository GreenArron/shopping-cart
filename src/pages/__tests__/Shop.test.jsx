import { useState } from "react";
import { it, expect, vi } from "vitest";
import { MemoryRouter, Routes, Route, Outlet } from "react-router-dom";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Shop from "../Shop";

function createMockApp(initialCart = {}) {
  function MockApp() {
    const [mockCart, setMockCart] = useState(initialCart);
    return (
      <div className="testing-mock">
        <Outlet context={[mockCart, setMockCart]} />
      </div>
    );
  }
  return <MockApp />;
}

async function renderCommon() {
  let renderedElement;

  await waitFor(() => {
    renderedElement = render(
      <MemoryRouter>
        <Routes location={"/products"}>
          <Route path="/" element={createMockApp()}>
            <Route path="/products" element={<Shop />} />
          </Route>
        </Routes>
      </MemoryRouter>,
    );
  });

  return renderedElement;
}

vi.mock("../../utils/fakestore-api", () => {
  const moddedModule = {
    fetchProducts: vi.fn(),
    fetchCategories: vi.fn(),
  };

  moddedModule.fetchCategories.mockResolvedValue([
    "show-test",
    "dontshow-test",
  ]);
  moddedModule.fetchProducts.mockResolvedValue([
    {
      id: 1,
      title: "TestProduct",
      price: 451,
      description: "Perfect mock test product",
      category: "show-test",
      image: "",
      rating: {
        rate: 5,
        count: 10,
      },
    },
  ]);
  return moddedModule;
});

it("matches last approved render", async () => {
  const { container } = await renderCommon();
  expect(container).toMatchSnapshot();
});

it("shows mocked test product", async () => {
  renderCommon();
  const product = await screen.findByRole("article");
  expect(product.textContent).toMatch(/TestProduct/i);
});

it("has sorter components", async () => {
  await renderCommon();
  expect(screen.getByRole("button", { name: /sort/i })).toBeDefined();
  expect(screen.getByRole("button", { name: /category/i })).toBeDefined();
});

it("correctly sorts categories", async () => {
  const user = userEvent.setup();
  renderCommon();

  const categoryButton = screen.getByRole("button", { name: /category/i });
  await user.click(categoryButton);

  const dontShowCategoryButton = screen.getByRole("button", {
    name: /dontshow-test/i,
  });
  await user.click(dontShowCategoryButton);
  expect(screen.queryByRole("article")).toBeNull();

  await user.click(categoryButton);
  const showCategoryButton = screen.getByRole("button", {
    name: /\bshow-test/i,
  });
  await user.click(showCategoryButton);
  expect(screen.getByRole("article")).toBeDefined();
});

// I can test the sort order and sort by here but god I don't need to ðŸ˜­
// THIS IS FINE!!!
