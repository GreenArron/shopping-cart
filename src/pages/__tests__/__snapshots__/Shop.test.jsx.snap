// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches last approved render 1`] = `
<div>
  <div
    class="testing-mock"
  >
    <main
      class="shop"
    >
      <section
        aria-label="filter controls"
        class="controls"
      >
        <div
          class="_filter-container_e8fa16"
        >
          <button>
            Category: All
          </button>
          <div
            aria-label="filter by category"
            class="_filter-content_e8fa16 false"
          />
        </div>
        <div
          class="sort-container"
        >
          <button
            class="sortorder disabled down"
          >
            <svg
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
              />
            </svg>
          </button>
          <div
            class="_filter-container_e8fa16"
          >
            <button>
              Sort By
            </button>
            <div
              aria-label="sort by"
              class="_filter-content_e8fa16 false"
            />
          </div>
        </div>
      </section>
      <section
        aria-label="shop items"
        class="products"
      >
        <article
          class="shop-product"
        >
          <img
            src=""
          />
          <div
            class="product-info"
          >
            <h3
              class="product-name"
            >
              TestProduct
            </h3>
            <div
              class="rating"
            >
              <div
                aria-label="rating: 5 out of 5"
                class="star-rating"
              >
                <span
                  class="full-star"
                >
                  <svg
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12,17.27L18.18,21L16.54,13.97L22,9.24L14.81,8.62L12,2L9.19,8.62L2,9.24L7.45,13.97L5.82,21L12,17.27Z"
                    />
                  </svg>
                </span>
                <span
                  class="full-star"
                >
                  <svg
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12,17.27L18.18,21L16.54,13.97L22,9.24L14.81,8.62L12,2L9.19,8.62L2,9.24L7.45,13.97L5.82,21L12,17.27Z"
                    />
                  </svg>
                </span>
                <span
                  class="full-star"
                >
                  <svg
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12,17.27L18.18,21L16.54,13.97L22,9.24L14.81,8.62L12,2L9.19,8.62L2,9.24L7.45,13.97L5.82,21L12,17.27Z"
                    />
                  </svg>
                </span>
                <span
                  class="full-star"
                >
                  <svg
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12,17.27L18.18,21L16.54,13.97L22,9.24L14.81,8.62L12,2L9.19,8.62L2,9.24L7.45,13.97L5.82,21L12,17.27Z"
                    />
                  </svg>
                </span>
                <span
                  class="full-star"
                >
                  <svg
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12,17.27L18.18,21L16.54,13.97L22,9.24L14.81,8.62L12,2L9.19,8.62L2,9.24L7.45,13.97L5.82,21L12,17.27Z"
                    />
                  </svg>
                </span>
              </div>
              <span
                class="rate-count"
              >
                (
                10
                )
              </span>
            </div>
            <div
              class="price"
            >
              $ 
              451
            </div>
          </div>
          <div
            class="buy-controls"
          >
            <button
              aria-description="adds this item to cart"
              class="cart-add"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 20C19 21.11 18.11 22 17 22C15.89 22 15 21.1 15 20C15 18.89 15.89 18 17 18C18.11 18 19 18.9 19 20M7 18C5.89 18 5 18.89 5 20C5 21.1 5.89 22 7 22C8.11 22 9 21.11 9 20S8.11 18 7 18M7.2 14.63L7.17 14.75C7.17 14.89 7.28 15 7.42 15H19V17H7C5.89 17 5 16.1 5 15C5 14.65 5.09 14.32 5.24 14.04L6.6 11.59L3 4H1V2H4.27L5.21 4H20C20.55 4 21 4.45 21 5C21 5.17 20.95 5.34 20.88 5.5L17.3 11.97C16.96 12.58 16.3 13 15.55 13H8.1L7.2 14.63M8.5 11H10V9H7.56L8.5 11M11 9V11H14V9H11M14 8V6H11V8H14M17.11 9H15V11H16L17.11 9M18.78 6H15V8H17.67L18.78 6M6.14 6L7.08 8H10V6H6.14Z"
                />
              </svg>
               
              <span
                aria-description="adds item to cart"
              >
                Add
              </span>
            </button>
          </div>
        </article>
      </section>
    </main>
  </div>
</div>
`;
